From 8f09530f6c1eb013374e6cd04b6d4b4633e6c333 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:39 -0400
Subject: [virt] kvm/ppc: book3s hv - Allow negative offsets to real-mode hcall handlers

Message-id: <1410545655-205645-111-git-send-email-dzickus@redhat.com>
Patchwork-id: 94120
O-Subject: [RHEL7 PATCH 110/626] KVM: PPC: Book3S HV: Allow negative offsets to real-mode hcall handlers
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 4baa1d871c8c959084aa5367a9bf211f383941e5
Author: Paul Mackerras <paulus@samba.org>
Date:   Mon Jul 8 20:09:53 2013 +1000

    KVM: PPC: Book3S HV: Allow negative offsets to real-mode hcall handlers

    The table of offsets to real-mode hcall handlers in book3s_hv_rmhandlers.S
    can contain negative values, if some of the handlers end up before the
    table in the vmlinux binary.  Thus we need to use a sign-extending load
    to read the values in the table rather than a zero-extending load.
    Without this, the host crashes when the guest does one of the hcalls
    with negative offsets, due to jumping to a bogus address.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index cb31e00..c71103b 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -1393,7 +1393,7 @@ hcall_try_real_mode:
  cmpldi r3,hcall_real_table_end - hcall_real_table
  bge guest_exit_cont
  LOAD_REG_ADDR(r4, hcall_real_table)
- lwzx r3,r3,r4
+ lwax r3,r3,r4
  cmpwi r3,0
  beq guest_exit_cont
  add r3,r3,r4
-- 
1.7.1